config/katello.migrations/200123161606-enable-pulpcore.rb

```diff
@@ -0,0 +1,5 @@
+cls = 'foreman_proxy::plugin::pulp'
+if answers[cls].is_a?(Hash)
+  answers[cls]['pulpcore_enabled'] = true
```

Comment:
This is the default so I don't think we need this if-condition. @ekohl ?

This migration will also need to set the following on upgrade:

```
proxy_pulp_isos_to_pulpcore: false
use_pulp_2_for_file: true
use_pulp_2_for_docker: true
```

Reply:
> This is the default so I don't think we need this if-condition. @ekohl ?

The code default is off: https://github.com/theforeman/puppet-foreman_proxy/blob/32b61e3f6a4ee150d65413c201b5cb902ad7ecd3/manifests/plugin/pulp/params.pp#L11

If you change supplied katello-answers.yaml it gives you a chance to detect upgrades.

> use_pulp_2_for_file: true
> use_pulp_2_for_docker: true

Note that these are on `katello`, not `foreman_proxy::plugin::pulp`.

Reply:
@ehelms @ekohl thanks.

> If you change supplied katello-answers.yaml it gives you a chance to detect upgrades.

@ekohl can you elaborate on this point a bit? what conditional can i use to detect upgrades and set the right params?

Reply:
Currently the answers looks like this:
```yaml
foreman_proxy::plugin::pulp: true
```

For users who already ran the installer this gets expanded to:
```yaml
foreman_proxy::plugin::pulp:
  enabled: true
  pulpnode_enabled: false
  # and others
```

I don't know for sure if the `pulpcore_enabled` default is already loaded during migrations, but probably not. If it is, it's `false` since that's the module default.

Now if you change the answers file to:
```yaml
foreman_proxy::plugin::pulp:
  pulpcore_enabled: true
```

This gives you the ability to distinguish a fresh install.

In the migration you can have the following cases that the installer can see and shouldn't crash on:
```ruby
answers['foreman_proxy::plugin::pulp'] == false # user turned off the module
answers['foreman_proxy::plugin::pulp'] == true # default in a previous version
answers['foreman_proxy::plugin::pulp'].is_a?(Hash) == true # the following cases imply this
answers['foreman_proxy::plugin::pulp']['pulpcore_enabled'] == nil # upgrading user
answers['foreman_proxy::plugin::pulp']['pulpcore_enabled'] == false # module default
answers['foreman_proxy::plugin::pulp']['pulpcore_enabled'] == true # fresh install
```

That makes me think the following should safely cover all cases:
```ruby
puppet_class = 'foreman_proxy::plugin::pulp'
param = 'pulpcore_enabled'

def upgrade
  answers[puppet_class][param] = true
  return unless answers['katello']
  answers['katello'] = {} unless answers['katello'].is_a?(Hash)
  answers['katello']['use_pulp_2_for_file'] = true
  answers['katello']['use_pulp_2_for_docker'] = true
end

if answers[puppet_class].is_a?(Hash)
  upgrade unless answers[puppet_class][param]
elsif answers[puppet_class]
  answers[puppet_class][param] = {}
  upgrade
end
```

You can see it ignores the fresh install which we can now safely detect. If the module is on in some way, we upgrade.

Writing this makes me think we need a migration helper in kafo:
```ruby
def set_answer(puppet_class, parameter, value)
  answers[puppet_class] = {} unless answers[puppet_class].is_a?(Hash)
  answers[puppet_class][parameter] = value
end
```

Reply:
@ekohl a big, sincere thanks for the detailed breakdown and recommendations.